In [ ]:
import pandas as pd

df = pd.read_csv('my_dataset_features.csv')
display(df.head())

mfcc_0      mfcc_1     mfcc_2     mfcc_3     mfcc_4     mfcc_5  \
0 -180.261490  123.208557  12.179420  25.168259   1.716785   8.663729   
1 -226.138382   52.582932  -0.217220  18.274734 -12.700911 -10.182571   
2 -294.740601  118.139587   9.248292   6.061491  -0.304913 -13.041572   
3 -141.580521   82.594872  12.024048  37.845757   5.738265   4.515667   
4 -209.824310  126.672104  -6.879056  31.562664   3.401516  -5.088613   

      mfcc_6     mfcc_7     mfcc_8     mfcc_9    mfcc_10   mfcc_11    mfcc_12  \
0   7.846949  19.999519   4.124840   5.814000  -5.934352 -2.364329 -12.630212   
1 -25.913273 -14.967414 -28.487017 -17.641758 -18.253475 -1.197921  -7.264189   
2 -16.138256 -11.038885  -7.435064  -6.881006  -8.622358 -4.305947  -1.445444   
3   6.479244   8.703977 -11.235284  16.566845  -0.778718  7.354715   1.187736   
4  10.778674   0.985945   1.518578  -3.790230  -5.671856  6.887865   0.264488   

      centroid    bandwidth      rolloff       zcr         label  
0  1342.833283  1905.003255  2562.851169  0.054439  melody songs  
1  2378.633113  2516.079719  5230.426706  0.090856  melody songs  
2  1213.902899  1424.499808  2060.904205  0.058951  melody songs  
3  2466.963081  2633.124881  5545.508134  0.104271  melody songs  
4  1424.919412  1702.190825  2689.642069  0.066759  melody songs

In [ ]:
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split

X = df.drop('label', axis=1)
y = df['label']

encoder = LabelEncoder()
y_encoded = encoder.fit_transform(y)

X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=0.2, random_state=42)

display(X_train.head())
display(y_train[0:5])

mfcc_0      mfcc_1     mfcc_2     mfcc_3     mfcc_4     mfcc_5  \
6  -149.316452  123.831635 -29.739553  27.770735  -2.377445 -10.376748   
26 -124.387581   95.315193 -25.463434  -1.714854 -23.422338  -8.222850   
12 -203.587845  103.690208   6.413342  29.019629   3.377941   0.478412   
17 -196.580322   72.587166  12.802266  23.369781   9.507194   2.529057   
9  -230.550690   99.653847  14.597300  22.789593   9.157195  14.030893   

       mfcc_6     mfcc_7     mfcc_8     mfcc_9    mfcc_10   mfcc_11   mfcc_12  \
6   -4.848413  -2.386041 -11.043736  -3.075054 -12.121479 -4.607691 -5.432042   
26 -10.481104   2.784271 -11.443886   4.876653  -1.011194  3.327785 -0.524326   
12 -13.478107  -9.229634  -8.881061  -0.817032  -6.338444 -0.993185  0.128573   
17   8.715775  -3.943071   6.488058 -10.632622 -14.003020 -7.098627 -2.793974   
9   -1.646349 -11.566245 -16.198318 -12.364244  -9.447881 -5.608509 -9.027602   

       centroid    bandwidth      rolloff       zcr  
6   1620.479305  1739.419300  3063.431476  0.081676  
26  2083.672296  2134.593182  4227.899434  0.092462  
12  1711.103337  2131.293708  3666.394496  0.064563  
17  2596.325686  2833.135526  5916.422558  0.090879  
9   1462.807366  1909.614851  3347.821423  0.038706

array([1, 0, 1, 1, 1])

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

model = Sequential()
model.add(Dense(64, input_shape=(X_train.shape[1],), activation='relu'))
model.add(Dense(32, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

model.summary()

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 64)             │         1,152 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,265 (12.75 KB)

 Trainable params: 3,265 (12.75 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
history = model.fit(X_train, y_train, epochs=50, batch_size=32, validation_split=0.2)

Epoch 1/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 5s 5s/step - accuracy: 0.4800 - loss: 75.3080 - val_accuracy: 0.5714 - val_loss: 40.9068
Epoch 2/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 217ms/step - accuracy: 0.4800 - loss: 40.3595 - val_accuracy: 0.2857 - val_loss: 32.9164
Epoch 3/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 149ms/step - accuracy: 0.4000 - loss: 23.2534 - val_accuracy: 0.1429 - val_loss: 54.0344
Epoch 4/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 328ms/step - accuracy: 0.4000 - loss: 32.3001 - val_accuracy: 0.2857 - val_loss: 60.1328
Epoch 5/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 142ms/step - accuracy: 0.4000 - loss: 38.1182 - val_accuracy: 0.2857 - val_loss: 54.8423
Epoch 6/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 151ms/step - accuracy: 0.4000 - loss: 34.4510 - val_accuracy: 0.1429 - val_loss: 43.0961
Epoch 7/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 178ms/step - accuracy: 0.4000 - loss: 25.0740 - val_accuracy: 0.1429 - val_loss: 27.5726
Epoch 8/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 161ms/step - accuracy: 0.3600 - loss: 16.9348 - val_accuracy: 0.5714

In [ ]:
evaluation_results = model.evaluate(X_test, y_test)
print(f"Test Loss: {evaluation_results[0]:.4f}")
print(f"Test Accuracy: {evaluation_results[1]:.4f}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 61ms/step - accuracy: 0.6667 - loss: 1.2704
Test Loss: 1.2704
Test Accuracy: 0.6667


In [ ]:
import librosa
import numpy as np
import pandas as pd

audio_path = '/content/WhatsApp Audio 2025-09-30 at 12.59.31_b9c6045f.mp3'
y_audio, sr = librosa.load(audio_path)

mfccs = librosa.feature.mfcc(y=y_audio, sr=sr, n_mfcc=13)
mfccs_mean = np.mean(mfccs.T, axis=0)

spectral_centroid = np.mean(librosa.feature.spectral_centroid(y=y_audio, sr=sr))
spectral_bandwidth = np.mean(librosa.feature.spectral_bandwidth(y=y_audio, sr=sr))
spectral_rolloff = np.mean(librosa.feature.spectral_rolloff(y=y_audio, sr=sr))
zero_crossing_rate = np.mean(librosa.feature.zero_crossing_rate(y=y_audio))

new_song_features = np.hstack((mfccs_mean, spectral_centroid, spectral_bandwidth, spectral_rolloff, zero_crossing_rate))

feature_names = [f'mfcc_{i}' for i in range(13)] + ['centroid', 'bandwidth', 'rolloff', 'zcr']
new_song_df = pd.DataFrame([new_song_features], columns=feature_names)

prediction = model.predict(new_song_df)
predicted_class = (prediction > 0.5).astype("int32")

predicted_genre = encoder.inverse_transform(predicted_class)

print(f"The predicted genre for the new song is: {predicted_genre[0]}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 80ms/step
The predicted genre for the new song is: Rap song


/usr/local/lib/python3.12/dist-packages/sklearn/preprocessing/_label.py:151: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
import pandas as pd

df = pd.read_csv('my_dataset_features.csv')
display(df.head())

mfcc_0      mfcc_1     mfcc_2     mfcc_3     mfcc_4     mfcc_5  \
0 -180.261490  123.208557  12.179420  25.168259   1.716785   8.663729   
1 -226.138382   52.582932  -0.217220  18.274734 -12.700911 -10.182571   
2 -294.740601  118.139587   9.248292   6.061491  -0.304913 -13.041572   
3 -141.580521   82.594872  12.024048  37.845757   5.738265   4.515667   
4 -209.824310  126.672104  -6.879056  31.562664   3.401516  -5.088613   

      mfcc_6     mfcc_7     mfcc_8     mfcc_9    mfcc_10   mfcc_11    mfcc_12  \
0   7.846949  19.999519   4.124840   5.814000  -5.934352 -2.364329 -12.630212   
1 -25.913273 -14.967414 -28.487017 -17.641758 -18.253475 -1.197921  -7.264189   
2 -16.138256 -11.038885  -7.435064  -6.881006  -8.622358 -4.305947  -1.445444   
3   6.479244   8.703977 -11.235284  16.566845  -0.778718  7.354715   1.187736   
4  10.778674   0.985945   1.518578  -3.790230  -5.671856  6.887865   0.264488   

      centroid    bandwidth      rolloff       zcr         label  
0  1342.833283  1905.003255  2562.851169  0.054439  melody songs  
1  2378.633113  2516.079719  5230.426706  0.090856  melody songs  
2  1213.902899  1424.499808  2060.904205  0.058951  melody songs  
3  2466.963081  2633.124881  5545.508134  0.104271  melody songs  
4  1424.919412  1702.190825  2689.642069  0.066759  melody songs

In [ ]:
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split

X = df.drop('label', axis=1)
y = df['label']

encoder = LabelEncoder()
y_encoded = encoder.fit_transform(y)

X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=0.2, random_state=42)

display(X_train.head())
display(y_train[0:5])

mfcc_0      mfcc_1     mfcc_2     mfcc_3     mfcc_4     mfcc_5  \
6  -149.316452  123.831635 -29.739553  27.770735  -2.377445 -10.376748   
26 -124.387581   95.315193 -25.463434  -1.714854 -23.422338  -8.222850   
12 -203.587845  103.690208   6.413342  29.019629   3.377941   0.478412   
17 -196.580322   72.587166  12.802266  23.369781   9.507194   2.529057   
9  -230.550690   99.653847  14.597300  22.789593   9.157195  14.030893   

       mfcc_6     mfcc_7     mfcc_8     mfcc_9    mfcc_10   mfcc_11   mfcc_12  \
6   -4.848413  -2.386041 -11.043736  -3.075054 -12.121479 -4.607691 -5.432042   
26 -10.481104   2.784271 -11.443886   4.876653  -1.011194  3.327785 -0.524326   
12 -13.478107  -9.229634  -8.881061  -0.817032  -6.338444 -0.993185  0.128573   
17   8.715775  -3.943071   6.488058 -10.632622 -14.003020 -7.098627 -2.793974   
9   -1.646349 -11.566245 -16.198318 -12.364244  -9.447881 -5.608509 -9.027602   

       centroid    bandwidth      rolloff       zcr  
6   1620.479305  1739.419300  3063.431476  0.081676  
26  2083.672296  2134.593182  4227.899434  0.092462  
12  1711.103337  2131.293708  3666.394496  0.064563  
17  2596.325686  2833.135526  5916.422558  0.090879  
9   1462.807366  1909.614851  3347.821423  0.038706

array([1, 0, 1, 1, 1])

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Conv1D, MaxPooling1D, Flatten

cnn_model = Sequential()
cnn_model.add(Conv1D(filters=64, kernel_size=3, activation='relu', input_shape=(X_train.shape[1], 1)))
cnn_model.add(MaxPooling1D(pool_size=2))
cnn_model.add(Flatten())
cnn_model.add(Dense(32, activation='relu'))
cnn_model.add(Dense(1, activation='sigmoid'))

cnn_model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

cnn_model.summary()

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d (Conv1D)                 │ (None, 15, 64)         │           256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d (MaxPooling1D)    │ (None, 7, 64)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 448)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 32)             │        14,368 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 14,657 (57.25 KB)

 Trainable params: 14,657 (57.25 KB)

 Non-trainable params: 0 (0.00 B)

## Reshape the input data

### Subtask:
Reshape the `X_train` and `X_test` data to have a 3D shape (samples, timesteps, features) as required by the Conv1D layers.


**Reasoning**:
Reshape X_train and X_test to be compatible with the Conv1D layer.



In [ ]:
X_train = X_train.values.reshape(X_train.shape[0], X_train.shape[1], 1)
X_test = X_test.values.reshape(X_test.shape[0], X_test.shape[1], 1)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

X_train shape: (32, 17, 1)
X_test shape: (9, 17, 1)


## Compile the cnn model

### Subtask:
Compile the new CNN model with an appropriate optimizer and loss function.


**Reasoning**:
Compile the new CNN model with the specified optimizer, loss function, and metric.



In [ ]:
cnn_model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

## Train the cnn model

### Subtask:
Train the CNN model using the reshaped training data.


**Reasoning**:
Train the compiled CNN model using the reshaped training data.



In [ ]:
history = cnn_model.fit(X_train, y_train, epochs=50, batch_size=32, validation_split=0.2)

Epoch 1/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 2s 2s/step - accuracy: 0.5200 - loss: 264.2656 - val_accuracy: 0.4286 - val_loss: 268.8327
Epoch 2/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 86ms/step - accuracy: 0.5200 - loss: 195.1286 - val_accuracy: 0.4286 - val_loss: 182.4261
Epoch 3/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 86ms/step - accuracy: 0.5200 - loss: 131.5343 - val_accuracy: 0.4286 - val_loss: 101.4105
Epoch 4/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 76ms/step - accuracy: 0.5200 - loss: 71.4398 - val_accuracy: 0.4286 - val_loss: 27.3028
Epoch 5/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 77ms/step - accuracy: 0.3600 - loss: 18.2648 - val_accuracy: 0.5714 - val_loss: 31.0257
Epoch 6/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 77ms/step - accuracy: 0.4800 - loss: 38.8904 - val_accuracy: 0.5714 - val_loss: 50.4096
Epoch 7/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 77ms/step - accuracy: 0.4800 - loss: 63.4206 - val_accuracy: 0.5714 - val_loss: 57.4172
Epoch 8/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 93ms/step - accuracy: 0.4800 - loss: 71.8654 - val_accuracy: 0.5714 

## Evaluate the cnn model

### Subtask:
Evaluate the performance of the trained CNN model on the reshaped test data.


**Reasoning**:
Evaluate the trained CNN model on the test data to assess its performance.



In [ ]:
evaluation_results_cnn = cnn_model.evaluate(X_test, y_test)
print(f"Test Loss (CNN): {evaluation_results_cnn[0]:.4f}")
print(f"Test Accuracy (CNN): {evaluation_results_cnn[1]:.4f}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 109ms/step - accuracy: 0.7778 - loss: 1.4561
Test Loss (CNN): 1.4561
Test Accuracy (CNN): 0.7778


## Update the prediction code

### Subtask:
Adjust the code for predicting the genre of a new song to use the reshaped input data for the CNN model.


**Reasoning**:
Reshape the new song features DataFrame to match the input shape of the CNN model (1, timesteps, features) and then use the trained CNN model to predict the genre, inverse transform the prediction to get the genre label, and print the result.



In [ ]:
# Reshape the new song features for CNN prediction
new_song_features_reshaped = new_song_df.values.reshape(new_song_df.shape[0], new_song_df.shape[1], 1)

# Predict the genre using the trained CNN model
prediction_cnn = cnn_model.predict(new_song_features_reshaped)

# Apply the threshold (0.5) to get the predicted class (0 or 1)
predicted_class_cnn = (prediction_cnn > 0.5).astype("int32")

# Use the encoder to transform the predicted class back to the original genre label
predicted_genre_cnn = encoder.inverse_transform(predicted_class_cnn)

# Print the predicted genre for the new song
print(f"The predicted genre for the new song using CNN is: {predicted_genre_cnn[0]}")

NameError: name 'new_song_df' is not defined

**Reasoning**:
The previous command failed because `new_song_df` was not defined. I need to regenerate the `new_song_df` DataFrame by loading the audio file and extracting its features, then proceed with reshaping and predicting using the CNN model.



In [ ]:
import librosa
import numpy as np
import pandas as pd

audio_path = '/content/WhatsApp Audio 2025-09-30 at 12.59.31_b9c6045f.mp3'
y_audio, sr = librosa.load(audio_path)

mfccs = librosa.feature.mfcc(y=y_audio, sr=sr, n_mfcc=13)
mfccs_mean = np.mean(mfccs.T, axis=0)

spectral_centroid = np.mean(librosa.feature.spectral_centroid(y=y_audio, sr=sr))
spectral_bandwidth = np.mean(librosa.feature.spectral_bandwidth(y=y_audio, sr=sr))
spectral_rolloff = np.mean(librosa.feature.spectral_rolloff(y=y_audio, sr=sr))
zero_crossing_rate = np.mean(librosa.feature.zero_crossing_rate(y=y_audio))

new_song_features = np.hstack((mfccs_mean, spectral_centroid, spectral_bandwidth, spectral_rolloff, zero_crossing_rate))

feature_names = [f'mfcc_{i}' for i in range(13)] + ['centroid', 'bandwidth', 'rolloff', 'zcr']
new_song_df = pd.DataFrame([new_song_features], columns=feature_names)

# Reshape the new song features for CNN prediction
new_song_features_reshaped = new_song_df.values.reshape(new_song_df.shape[0], new_song_df.shape[1], 1)

# Predict the genre using the trained CNN model
prediction_cnn = cnn_model.predict(new_song_features_reshaped)

# Apply the threshold (0.5) to get the predicted class (0 or 1)
predicted_class_cnn = (prediction_cnn > 0.5).astype("int32")

# Use the encoder to transform the predicted class back to the original genre label
predicted_genre_cnn = encoder.inverse_transform(predicted_class_cnn)

# Print the predicted genre for the new song
print(f"The predicted genre for the new song using CNN is: {predicted_genre_cnn[0]}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 75ms/step
The predicted genre for the new song using CNN is: Rap song


/usr/local/lib/python3.12/dist-packages/sklearn/preprocessing/_label.py:151: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


## Summary:

### Data Analysis Key Findings

*   The CNN model achieved a test accuracy of approximately 0.7778.
*   The predicted genre for the new song using the trained CNN model is "Rap song".

### Insights or Next Steps

*   The CNN model showed promising performance on the test data. Further hyperparameter tuning and exploring different CNN architectures could potentially improve accuracy.
*   Consider evaluating the model using additional metrics beyond accuracy, such as precision, recall, and F1-score, to get a more comprehensive understanding of its performance.
